File: templates/Uwp/_composition/_shared/Page.Ink.AddBasicInkServices/Services/Ink/InkStrokesService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Numerics;
using System.Threading.Tasks;
using Windows.Foundation;
using Windows.Storage;
using Windows.Storage.Provider;
using Windows.UI.Input.Inking;
using Windows.UI.Input.Inking.Analysis;
using Windows.UI.Xaml.Media;
using Param_ItemNamespace.EventHandlers.Ink;

namespace Param_ItemNamespace.Services.Ink
{
    public class InkStrokesService
    {
        private readonly InkStrokeContainer _strokeContainer;

        public InkStrokesService(InkStrokeContainer strokeContainer)
        {
            _strokeContainer = strokeContainer;
        }

        public event EventHandler<AddStrokeEventArgs> AddStrokeEvent;

        public event EventHandler<RemoveEventArgs> RemoveStrokeEvent;

        public event EventHandler<MoveStrokesEventArgs> MoveStrokesEvent;

        public event EventHandler<EventArgs> ClearStrokesEvent;

        public event EventHandler<CopyPasteStrokesEventArgs> CutStrokesEvent;

        public event EventHandler<CopyPasteStrokesEventArgs> PasteStrokesEvent;

        public event EventHandler<EventArgs> LoadInkFileEvent;

        public InkStroke AddStroke(InkStroke stroke)
        {
            var newStroke = stroke.Clone();
            _strokeContainer.AddStroke(newStroke);

            AddStrokeEvent?.Invoke(this, new AddStrokeEventArgs(newStroke, stroke));

            return newStroke;
        }

        public bool RemoveStroke(InkStroke stroke)
        {
            var deleteStroke = GetStrokes().FirstOrDefault(s => s.Id == stroke.Id);
            if (deleteStroke == null)
            {
                return false;
            }

            ClearStrokesSelection();
            deleteStroke.Selected = true;
            _strokeContainer.DeleteSelected();

            RemoveStrokeEvent?.Invoke(this, new RemoveEventArgs(stroke));
            return true;
        }

        public bool RemoveStrokesByIds(IEnumerable<uint> strokeIds)
        {
            var strokes = GetStrokesByIds(strokeIds);

            foreach (var stroke in strokes)
            {
                RemoveStroke(stroke);
            }

            return strokes.Any();
        }

        public IEnumerable<InkStroke> GetStrokes() => _strokeContainer.GetStrokes();

        public IEnumerable<InkStroke> GetSelectedStrokes() => GetStrokes().Where(s => s.Selected);

        public void ClearStrokes()
        {
            _strokeContainer.Clear();
            ClearStrokesEvent?.Invoke(this, EventArgs.Empty);
        }

        public void ClearStrokesSelection()
        {
            foreach (var stroke in GetStrokes())
            {
                stroke.Selected = false;
            }
        }

        public Rect SelectStrokes(IEnumerable<InkStroke> strokes)
        {
            ClearStrokesSelection();

            foreach (var stroke in strokes)
            {
                stroke.Selected = true;
            }

            return GetRectBySelectedStrokes();
        }

        public Rect SelectStrokesByNode(IInkAnalysisNode node)
        {
            var ids = GetNodeStrokeIds(node);
            var strokes = GetStrokesByIds(ids);
            var rect = SelectStrokes(strokes);

            return rect;
        }

        public Rect SelectStrokesByPoints(PointCollection points)
        {
            ClearStrokesSelection();
            return _strokeContainer.SelectWithPolyLine(points);
        }

        public void MoveSelectedStrokes(Point startPosition, Point endPosition)
        {
            var x = (float)(endPosition.X - startPosition.X);
            var y = (float)(endPosition.Y - startPosition.Y);

            var matrix = Matrix3x2.CreateTranslation(x, y);

            if (!matrix.IsIdentity)
            {
                var selectedStrokes = GetSelectedStrokes();
                foreach (var stroke in selectedStrokes)
                {
                    stroke.PointTransform *= matrix;
                }

                MoveStrokesEvent?.Invoke(this, new MoveStrokesEventArgs(selectedStrokes, startPosition, endPosition));
            }
        }

        public Rect CopySelectedStrokes()
        {
            _strokeContainer.CopySelectedToClipboard();
            return GetRectBySelectedStrokes();
        }

        public Rect CutSelectedStrokes()
        {
            var rect = CopySelectedStrokes();

            var selectedStrokes = GetSelectedStrokes().ToList();

            foreach (var stroke in selectedStrokes)
            {
                RemoveStroke(stroke);
            }

            CutStrokesEvent?.Invoke(this, new CopyPasteStrokesEventArgs(selectedStrokes));

            return rect;
        }

        public Rect PasteSelectedStrokes(Point position)
        {
            var rect = Rect.Empty;

            if (CanPaste)
            {
                var ids = GetStrokes().Select(s => s.Id).ToList();

                rect = _strokeContainer.PasteFromClipboard(position);

                var pastedStrokes = _strokeContainer.GetStrokes().Where(s => !ids.Contains(s.Id));

                PasteStrokesEvent?.Invoke(this, new CopyPasteStrokesEventArgs(pastedStrokes));
            }

            return rect;
        }

        public bool CanPaste => _strokeContainer.CanPasteFromClipboard();

        public async Task<bool> LoadInkFileAsync(StorageFile file)
        {
            if (file == null)
            {
                return false;
            }

            ClearStrokesSelection();
            ClearStrokes();

            using (var stream = await file.OpenSequentialReadAsync())
            {
                await _strokeContainer.LoadAsync(stream);
            }

            LoadInkFileEvent?.Invoke(this, EventArgs.Empty);
            return true;
        }

        public async Task<FileUpdateStatus> SaveInkFileAsync(StorageFile file)
        {
            if (file != null)
            {
                // Prevent updates to the file until updates are finalized with call to CompleteUpdatesAsync.
                CachedFileManager.DeferUpdates(file);

                using (var stream = await file.OpenAsync(FileAccessMode.ReadWrite))
                {
                    await _strokeContainer.SaveAsync(stream);
                }

                // Finalize write so other apps can update file.
                return await CachedFileManager.CompleteUpdatesAsync(file);
            }

            return FileUpdateStatus.Failed;
        }

        private IEnumerable<InkStroke> GetStrokesByIds(IEnumerable<uint> strokeIds)
        {
            foreach (var strokeId in strokeIds)
            {
                yield return _strokeContainer.GetStrokeById(strokeId);
            }
        }

        private IReadOnlyList<uint> GetNodeStrokeIds(IInkAnalysisNode node)
        {
            var strokeIds = node.GetStrokeIds();
            if (node.Kind == InkAnalysisNodeKind.Paragraph && node.Children[0].Kind == InkAnalysisNodeKind.ListItem)
            {
                strokeIds = new HashSet<uint>(strokeIds).ToList();
            }

            return strokeIds;
        }

        private Rect GetRectBySelectedStrokes()
        {
            var rect = Rect.Empty;
            foreach (var stroke in GetSelectedStrokes())
            {
                rect.Union(stroke.BoundingRect);
            }

            return rect;
        }
    }
}

File: templates/Uwp/_composition/_shared/Page.Ink.AddBasicInkServices._VB/Services/Ink/InkStrokesService.vb
Imports System.Numerics
Imports Windows.Storage
Imports Windows.Storage.Provider
Imports Windows.UI.Input.Inking
Imports Windows.UI.Input.Inking.Analysis
Imports Param_RootNamespace.EventHandlers.Ink

Namespace Services.Ink
    Public Class InkStrokesService
        Private ReadOnly _strokeContainer As InkStrokeContainer

        Public Sub New(inkPresenter As InkPresenter)
            _strokeContainer = inkPresenter.StrokeContainer
            AddHandler inkPresenter.StrokesCollected, Sub(s, e) RaiseEvent StrokesCollected(Me, e)
            AddHandler inkPresenter.StrokesErased, Sub(s, e) RaiseEvent StrokesErased(Me, e)
        End Sub

        Public Event StrokesCollected As EventHandler(Of InkStrokesCollectedEventArgs)

        Public Event StrokesErased As EventHandler(Of InkStrokesErasedEventArgs)

        Public Event AddStrokeEvent As EventHandler(Of AddStrokeEventArgs)

        Public Event RemoveStrokeEvent As EventHandler(Of RemoveEventArgs)

        Public Event MoveStrokesEvent As EventHandler(Of MoveStrokesEventArgs)

        Public Event ClearStrokesEvent As EventHandler(Of EventArgs)

        Public Event CopyStrokesEvent As EventHandler(Of CopyPasteStrokesEventArgs)

        Public Event CutStrokesEvent As EventHandler(Of CopyPasteStrokesEventArgs)

        Public Event PasteStrokesEvent As EventHandler(Of CopyPasteStrokesEventArgs)

        Public Event LoadInkFileEvent As EventHandler(Of EventArgs)

        Public Event SelectStrokesEvent As EventHandler(Of EventArgs)

        Public Function AddStroke(stroke As InkStroke) As InkStroke
            Dim newStroke = stroke.Clone()
            _strokeContainer.AddStroke(newStroke)
            RaiseEvent AddStrokeEvent(Me, New AddStrokeEventArgs(newStroke, stroke))
            Return newStroke
        End Function

        Public Function RemoveStroke(stroke As InkStroke) As Boolean
            Dim deleteStroke = GetStrokes().FirstOrDefault(Function(s) s.Id = stroke.Id)

            If deleteStroke Is Nothing Then
                Return False
            End If

            ClearStrokesSelection()
            deleteStroke.Selected = True
            _strokeContainer.DeleteSelected()
            RaiseEvent RemoveStrokeEvent(Me, New RemoveEventArgs(stroke))
            Return True
        End Function

        Public Function RemoveStrokesByIds(strokeIds As IEnumerable(Of UInteger)) As Boolean
            Dim strokes = GetStrokesByIds(strokeIds)

            For Each stroke In strokes
                RemoveStroke(stroke)
            Next

            Return strokes.Any()
        End Function

        Public Function GetStrokes() As IEnumerable(Of InkStroke)
            Return _strokeContainer.GetStrokes()
        End Function

        Public Function GetSelectedStrokes() As IEnumerable(Of InkStroke)
            Return GetStrokes().Where(Function(stroke) stroke.Selected)
        End Function

        Public Sub ClearStrokes()
            _strokeContainer.Clear()
            RaiseEvent ClearStrokesEvent(Me, EventArgs.Empty)
        End Sub

        Public Sub ClearStrokesSelection()
            For Each stroke In GetStrokes()
                stroke.Selected = False
            Next
            RaiseEvent SelectStrokesEvent(Me, EventArgs.Empty)
        End Sub

        Public Function SelectStrokes(strokes As IEnumerable(Of InkStroke)) As Rect
            ClearStrokesSelection()

            For Each stroke In strokes
                stroke.Selected = True
            Next

            RaiseEvent SelectStrokesEvent(Me, EventArgs.Empty)
            Return GetRectBySelectedStrokes()
        End Function

        Public Function SelectStrokesByNode(node As IInkAnalysisNode) As Rect
            Dim ids = GetNodeStrokeIds(node)
            Dim strokes = GetStrokesByIds(ids)
            Dim rect = SelectStrokes(strokes)
            Return rect
        End Function

        Public Function SelectStrokesByPoints(points As PointCollection) As Rect
            ClearStrokesSelection()
            Dim rect = _strokeContainer.SelectWithPolyLine(points)
            RaiseEvent SelectStrokesEvent(Me, EventArgs.Empty)
            Return rect
        End Function

        Public Sub MoveSelectedStrokes(startPosition As Point, endPosition As Point)
            Dim x = CSng((endPosition.X - startPosition.X))
            Dim y = CSng((endPosition.Y - startPosition.Y))
            Dim matrix = Matrix3x2.CreateTranslation(x, y)

            If Not matrix.IsIdentity Then
                Dim selectedStrokes = GetSelectedStrokes()

                For Each stroke In selectedStrokes
                    stroke.PointTransform *= matrix
                Next

                RaiseEvent MoveStrokesEvent(Me, New MoveStrokesEventArgs(selectedStrokes, startPosition, endPosition))
            End If
        End Sub

        Public Function CopySelectedStrokes() As Rect
            _strokeContainer.CopySelectedToClipboard()
            RaiseEvent CopyStrokesEvent(Me, New CopyPasteStrokesEventArgs(GetSelectedStrokes()))
            Return GetRectBySelectedStrokes()
        End Function

        Public Function CutSelectedStrokes() As Rect
            Dim rect = CopySelectedStrokes()
            Dim selectedStrokes = GetSelectedStrokes().ToList()

            For Each stroke In selectedStrokes
                RemoveStroke(stroke)
            Next

            RaiseEvent CutStrokesEvent(Me, New CopyPasteStrokesEventArgs(selectedStrokes))
            Return rect
        End Function

        Public Function PasteSelectedStrokes(position As Point) As Rect
            Dim rect = Windows.Foundation.Rect.Empty

            If CanPaste Then
                Dim ids = GetStrokes().[Select](Function(s) s.Id).ToList()
                rect = _strokeContainer.PasteFromClipboard(position)
                Dim pastedStrokes = _strokeContainer.GetStrokes().Where(Function(stroke) Not ids.Contains(stroke.Id))
                RaiseEvent PasteStrokesEvent(Me, New CopyPasteStrokesEventArgs(pastedStrokes))
            End If

            Return rect
        End Function

        Public ReadOnly Property CanPaste As Boolean
            Get
                Return _strokeContainer.CanPasteFromClipboard()
            End Get
        End Property

        Public Async Function LoadInkFileAsync(file As StorageFile) As Task(Of Boolean)
            If file Is Nothing Then
                Return False
            End If

            ClearStrokesSelection()
            ClearStrokes()

            Using stream = Await file.OpenSequentialReadAsync()
                Await _strokeContainer.LoadAsync(stream)
            End Using

            RaiseEvent LoadInkFileEvent(Me, EventArgs.Empty)
            Return True
        End Function

        Public Async Function SaveInkFileAsync(file As StorageFile) As Task(Of FileUpdateStatus)
            If file IsNot Nothing Then
                ' Prevent updates to the file until updates are finalized with call to CompleteUpdatesAsync.
                CachedFileManager.DeferUpdates(file)

                Using stream = Await file.OpenAsync(FileAccessMode.ReadWrite)
                    Await _strokeContainer.SaveAsync(stream)
                End Using

                ' Finalize write so other apps can update file.
                Return Await CachedFileManager.CompleteUpdatesAsync(file)
            End If

            Return FileUpdateStatus.Failed
        End Function

        Private Iterator Function GetStrokesByIds(strokeIds As IEnumerable(Of UInteger)) As IEnumerable(Of InkStroke)
            For Each strokeId In strokeIds
                Yield _strokeContainer.GetStrokeById(strokeId)
            Next
        End Function

        Private Function GetNodeStrokeIds(node As IInkAnalysisNode) As IReadOnlyList(Of UInteger)
            Dim strokeIds = node.GetStrokeIds()

            If node.Kind = InkAnalysisNodeKind.Paragraph AndAlso node.Children(0).Kind = InkAnalysisNodeKind.ListItem Then
                strokeIds = New HashSet(Of UInteger)(strokeIds).ToList()
            End If

            Return strokeIds
        End Function

        Private Function GetRectBySelectedStrokes() As Rect
            Dim rect = Windows.Foundation.Rect.Empty

            For Each stroke In GetSelectedStrokes()
                rect.Union(stroke.BoundingRect)
            Next

            Return rect
        End Function
    End Class
End Namespace
